enhancements/kube-apiserver/stability-point-to-point-network-check.md

```diff
@@ -0,0 +1,447 @@
+
+### Non-Goals
+
+## Proposal
+
+### 1. One instance of a new `PodNetworkConnectivityCheck` CR per point to point connection.
+
+A point to point connection is between a source pod and a target endpoint.
+
+Component operators will create a `PodNetworkConnectivityCheck` resource per point-to-point connection. If an operator 
+is not managing the source pod directly, but via a workload controller (such as via a deployment or daemonset), the 
+operator will observe the workload resources to determine the list of pods and create a `PodNetworkConnectivityCheck` 
+resource for each target endpoint per pod. Operators that directly manage a source pod (such as operators that manage 
+static-pods) will create `PodNetworkConnectivityCheck` resources for each target endpoint when creating the source pod.
+
+### 2. Detailed status, probably some kind of latency information.
+
+The `PodNetworkConnectivityCheck` will maintain a log of the actions performed during a check, recording if the actions
+were successful, a reason for success or failure, latency of performing the action and some human readable message.
+
+**LogEntry:**
+```go
+// LogEntry records events
+type LogEntry struct {
+	// Start time of check action.
+	Start metav1.Time `json:"time,omitempty"`
+	// Status indicates if the action performed by the check was successful or not.
+	Success bool `json:"status"`
+	// Reason for status in a machine readable format.
+	Reason LogEntryReason `json:"reason"`
+	// Message explaining status in a human readable format.
+	Message string `json:"message"`
+	// Latency records how long the action mentioned in the entry took.
+	Latency time.Duration `json:"latency"`
+}
+``` 
+
+The start and end time of detected outages will be logged separately from the individual check actions.
+
+### 3. It should be possible to clean up garbage after a couple hours of inactivity.
+
+A controller to prune `PodNetworkConnectivityCheck` resources, will be provided for inclusion into source pod operators.
+
+A `PodNetworkConnectivityCheck` resource would be pruned after 48 hours of inactivity.
+
+Tools updating `PodNetworkConnectivityCheck` log entries should simultaneously delete the oldest log entry to ensure
+there are never more that 20 entries.
+
+### 4. Don't delete aggressively, on upgrades, we have different endpoints
+
+Source pods managed by workload controllers will have new `PodNetworkConnectivityCheck` resources to go along with their
+new names. The old `PodNetworkConnectivityCheck` resource will eventually go stale and be cleaned up by the provided
+pruning controller. 
+
+Directly managed pods whose names do not change (such as static pods) will share `PodNetworkConnectivityCheck` resources
+across revisions and upgrades.
+
+### 5. Create a binary that can take multiple destinations paired with instances to write to.
+
+### 6. Use that binary in containers included in each kas and oas (or other interesting) pod.
+
+The binary will:
```

Comment:
I do not see why this is separate from the pod itself.  Why do you need separate things to do these checks and an api and a client side API?  Why is this not output as events.  You haven't clearly described why you need an out of process tool for this (vs some arbitrary "oh it's separate" which you know I don't put much weight on).  For instaenc, the contrast is you simply capture these errors in your pods and report them elsewhere.

I don't envision other components using these checks the way you have described it - it's both too specific to your problem (early launch connectivity) and too misaligned with monitoring, metrics, existing readiness or health checks on nodes.

Reply:
> do not see why this is separate from the pod itself

this proposes creating another container in an existing pod.



> You haven't clearly described why you need an out of process tool for this (vs some arbitrary "oh it's separate" which you know I don't put much weight on).

We know that we want this for multiple pods already.  It doesn't seem to make sense to duplicate logic in all the binaries we would want to do this in.  Particularly when we consider that we want to experiment first, see if it works and build from there.



> Why is this not output as events.

Sure, once we have a detection mechanism we can report it via all-the-things, but we want to work out that detection first with a cheap prototype.



> misaligned with monitoring, metrics, existing readiness or health checks on nodes

1. nothing prevents alignment to metrics. We aren't against metrics.  We want to show the metrics we're interested in and confirm they're the ones we're interested in and we're happy to de-dupe *after* we find what we're looking for.
2. this is definitely  *not* a readiness or health check. If we lose connectivity with a particular kube-apiserver service, it is almost *never* the correct answer to kill the kube-apiserver container and very rarely the correct answer to remove ourselves from the load balancer. That will only make the outage worse: going from a partial outage to an everything is down outage.

If we build out this enhance and can backfeed to see that one of three masters is bad, we could later build a "kill this kube-apiserver" logic into it.  I want the option, but I don't want to start there.

We're trying to gather information that we think is missing.  This isn't fighting with any other initiative, it's just a debugging tool that we want to vet because we think it would make many of our other bugs more shallow.